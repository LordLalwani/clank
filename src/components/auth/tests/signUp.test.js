import React from "react";
import {
  render,
  queryByAttribute,
  cleanup,
  fireEvent,
  waitFor,
} from "@testing-library/react";
import SignUp from "../signUp";

import Amplify, { Auth } from "aws-amplify";
import awsconfig from "../../../aws-exports";
import renderer from "react-test-renderer";

Amplify.configure(awsconfig);

const getById = queryByAttribute.bind(null, "id");
afterEach(cleanup);

let DOM;
let emailTextField;
let passwordTextField;
let confirmPasswordTextField;
let verifyEmailButton;
let confirmationCodeTextfield;
let confirmationCodeButton;

beforeEach(() => {
  DOM = render(<SignUp />);
  emailTextField = getById(DOM.container, "email-textfield");
  passwordTextField = getById(DOM.container, "password-textfield");
  confirmPasswordTextField = getById(
    DOM.container,
    "confirm-password-textfield"
  );
  verifyEmailButton = getById(DOM.container, "verify-email-button");
  confirmationCodeTextfield = getById(
    DOM.container,
    "confirmation-code-textfield"
  );
  confirmationCodeButton = getById(DOM.container, "confirmation-code-button");
});

describe("render tests", () => {
  it("should render all sub components correctly", () => {
    expect(emailTextField).toBeTruthy();
    expect(passwordTextField).toBeTruthy();
    expect(confirmPasswordTextField).toBeTruthy();
    expect(verifyEmailButton).toBeTruthy();
  });

  it("should render signUp component correctly", () => {
    const tree = renderer.create(<SignUp />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("Behavior tests", () => {
  it("should display 'passwords don't match' when the password and confirm password textfields are not equal", () => {
    expect(DOM.queryByText("Passwords don't match")).not.toBeTruthy();
    fireEvent.change(passwordTextField, { target: { value: "notTheSame" } });
    fireEvent.change(confirmPasswordTextField, {
      target: { value: "sameTheNot" },
    });
    expect(DOM.queryByText("Passwords don't match")).toBeTruthy();

    // checks that `passwords don't match` message doesn't show on inverse
    fireEvent.change(passwordTextField, { target: { value: "samesies" } });
    fireEvent.change(confirmPasswordTextField, {
      target: { value: "samesies" },
    });
    expect(DOM.queryByText("Passwords don't match")).not.toBeTruthy();
  });

  it("should display the code confirmation textfield and button upon clicking `verify my email`", async () => {
    Auth.signUp = jest.fn().mockImplementation(() => {
      const response = { user: "larryTheLama" };
      return response;
    });

    fireEvent.change(emailTextField, { target: { value: "teeest@test.com" } });
    fireEvent.change(passwordTextField, {
      target: { value: "samesies123456" },
    });
    fireEvent.change(confirmPasswordTextField, {
      target: { value: "samesies123456" },
    });
    expect(verifyEmailButton).not.toHaveAttribute("disabled");
    expect(confirmationCodeTextfield).not.toBeTruthy();
    expect(confirmationCodeButton).not.toBeTruthy();
    expect(DOM.queryByText("Confirm")).not.toBeTruthy();

    fireEvent.click(verifyEmailButton);

    await waitFor(() => {
      expect(Auth.signUp).toHaveBeenCalledTimes(1);
    });
    expect(DOM.queryByText("Confirm")).toBeTruthy();

    // Was having problems targeting confirmationCodeButton so queried by autoComplete
    // this should be replaced asap >:[
    const getByAutoComplete = queryByAttribute.bind(null, "autoComplete");
    const confirmationCodeTextField2 = getByAutoComplete(DOM.container, "code");
    expect(confirmationCodeTextField2).toBeTruthy();
  });

  it("should make 'verify my email' button not disabled given textfields are not empty and passwords match", () => {
    expect(verifyEmailButton).toHaveAttribute("disabled");
    fireEvent.change(emailTextField, { target: { value: "test@test.com" } });
    fireEvent.change(passwordTextField, {
      target: { value: "samesies123456" },
    });
    fireEvent.change(confirmPasswordTextField, {
      target: { value: "samesies123456" },
    });
    expect(verifyEmailButton).not.toHaveAttribute("disabled");
  });
});
