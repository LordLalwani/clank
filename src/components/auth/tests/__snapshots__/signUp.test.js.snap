// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render tests should render signUp component correctly 1`] = `
<div
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "height": "90vh",
    }
  }
>
  <main
    className="MuiContainer-root MuiContainer-maxWidthXs"
  >
    <div
      className="SignUp-paper-12"
    >
      <div
        className="MuiAvatar-root MuiAvatar-circle SignUp-avatar-13 MuiAvatar-colorDefault"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zM9 6c0-1.66 1.34-3 3-3s3 1.34 3 3v2H9V6zm9 14H6V10h12v10zm-6-3c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2z"
          />
        </svg>
      </div>
      <h1
        className="MuiTypography-root MuiTypography-h5"
      >
        Sign up to Clank
      </h1>
      <form
        className="SignUp-form-14"
        noValidate={true}
      >
        <div
          className="MuiFormControl-root MuiTextField-root SignUp-textField-15 MuiFormControl-marginNormal MuiFormControl-fullWidth"
          onClick={[Function]}
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
            data-shrink={false}
            htmlFor="email-textfield"
            id="email-textfield-label"
          >
            Email Address
            <span
              aria-hidden={true}
              className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoComplete="email"
              autoFocus={true}
              className="MuiInputBase-input MuiOutlinedInput-input"
              disabled={false}
              id="email-textfield"
              name="email"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={true}
              type="text"
            />
            <fieldset
              aria-hidden={true}
              className="PrivateNotchedOutline-root-17 MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="PrivateNotchedOutline-legendLabelled-19"
              >
                <span>
                  Email Address
                   *
                </span>
              </legend>
            </fieldset>
          </div>
          
        </div>
        <div
          className="MuiFormControl-root MuiTextField-root SignUp-textField-15 MuiFormControl-marginNormal MuiFormControl-fullWidth"
          onClick={[Function]}
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
            data-shrink={false}
            htmlFor="password-textfield"
            id="password-textfield-label"
          >
            Password
            <span
              aria-hidden={true}
              className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoComplete="current-password"
              autoFocus={false}
              className="MuiInputBase-input MuiOutlinedInput-input"
              disabled={false}
              id="password-textfield"
              name="password"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={true}
              type="password"
            />
            <fieldset
              aria-hidden={true}
              className="PrivateNotchedOutline-root-17 MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="PrivateNotchedOutline-legendLabelled-19"
              >
                <span>
                  Password
                   *
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          className="MuiFormControl-root MuiTextField-root SignUp-textField-15 MuiFormControl-marginNormal MuiFormControl-fullWidth"
          onClick={[Function]}
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
            data-shrink={false}
            htmlFor="confirm-password-textfield"
            id="confirm-password-textfield-label"
          >
            Confirm Password
            <span
              aria-hidden={true}
              className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoComplete="current-password"
              autoFocus={false}
              className="MuiInputBase-input MuiOutlinedInput-input"
              disabled={false}
              id="confirm-password-textfield"
              name="Confirm Password"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={true}
              type="password"
            />
            <fieldset
              aria-hidden={true}
              className="PrivateNotchedOutline-root-17 MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="PrivateNotchedOutline-legendLabelled-19"
              >
                <span>
                  Confirm Password
                   *
                </span>
              </legend>
            </fieldset>
          </div>
          
        </div>
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained Button-button-21 Mui-disabled MuiButton-fullWidth Mui-disabled"
          disabled={true}
          id="verify-email-button"
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          type="submit"
        >
          <span
            className="MuiButton-label"
          >
            Verify My Email
          </span>
        </button>
        <div
          className="MuiGrid-root MuiGrid-container"
        >
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true"
          >
            <a
              className="MuiTypography-root MuiLink-root MuiLink-underlineHover SignUp-link-16 MuiTypography-body2 MuiTypography-colorPrimary"
              href="/password-recovery"
              onBlur={[Function]}
              onFocus={[Function]}
            >
              Forgot password?
            </a>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item"
          >
            <a
              className="MuiTypography-root MuiLink-root MuiLink-underlineHover SignUp-link-16 MuiTypography-body2 MuiTypography-colorPrimary"
              href="/sign-in"
              onBlur={[Function]}
              onFocus={[Function]}
            >
              Sign in
            </a>
          </div>
        </div>
      </form>
    </div>
    <div
      className="MuiBox-root MuiBox-root-23"
    >
      <p
        className="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-alignCenter"
      >
        Copyright © 
        <a
          className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorInherit"
          href="https://material-ui.com/"
          onBlur={[Function]}
          onFocus={[Function]}
        >
          Clank
        </a>
         
        2020
        .
      </p>
    </div>
  </main>
</div>
`;
